com.unity.multiplayer.mlapi/Tests/Runtime/NetworkVariableTests.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Collections;
+using UnityEngine;
+using UnityEngine.TestTools;
+using NUnit.Framework;
+
+namespace MLAPI.RuntimeTests
+{
+    public class NetworkVariableTests
+    {
+        [SetUp]
+        public void Setup()
+        {
+            // Create, instantiate, and host
+            Assert.IsTrue(NetworkManagerHelper.StartNetworkManager(out _));
+        }
+
+        /// <summary>
+        /// Runs generalized tests on all predefined NetworkVariable types
+        /// </summary>
+        [UnityTest]
+        public IEnumerator TestAllNetworkVariableTypes()
+        {
+            Guid gameObjectId = NetworkManagerHelper.AddGameNetworkObject("NetworkVariableTestComponent");
+
+            var networkVariableTestComponent = NetworkManagerHelper.AddComponentToObject<NetworkVariableTestComponent>(gameObjectId);
+
+            NetworkManagerHelper.SpawnNetworkObject(gameObjectId);
+
+            // Start Testing
+            networkVariableTestComponent.EnableTesting = true;
+
+            var testsAreComplete = networkVariableTestComponent.IsTestComplete();
+
+            // Wait for the NetworkVariable tests to complete
+            while (!testsAreComplete)
+            {
+                yield return new WaitForSeconds(0.003f);
+                testsAreComplete = networkVariableTestComponent.IsTestComplete();
```

Comment:
I don't understand. `m_FinishedTests` seems to be set right after modifying the variables. Yet, it seems like we test right away that they changed. How do we account for latency? What if it took multiples frames for the network values to change? Is the test somehow relying on running in a test environment with no latency?

Reply:
After thinking a bit more. It seems like this relies on the OnValueChange immediate notification callback. Which will be removed when we add snapshot, as the local client should not learn right away of their action. So, I fear the test would break when we start on actual snapshot, or rather on the "proper notification" dependency that snapshot will provide.

Reply:
These are "generic tests" that don't cover a networked session.
This will be further expanded once we can do this in a unit test.

Reply:
ok, but it will still break when the local, immediate, notification is removed

Reply:
if we're not really testing `NetworkVariable` functionality in a networking context, I wonder what's the value add here apart from making code coverage look fancier? :)

Reply:
Well it does a few things:
- We now have a NetworkVariable test that can be updated once we do have a network session
- There is now a way to generically test the NetworkVariable changed via NetworkVariableHelper
- This also provides general coverage that will be informative if we break some form of the NetworkVariable "basic functionality" in the future.

So, I think the bottom line value add is that we are at least testing these now and have something to update as we move forward?

Reply:
Sure. Well, when we get to the point where we test NetworkVariables in multiplayer game session, I expect things to change here and there but maybe proving that API doesn't break and also having a base tests skeleton once we get there are the value adds I was after. It's fine, I was just curious if I was missing something.

Reply:
It is missing things...and stuff.
But...it is one small step for MLAPI and one giant leap for NetworkVariable unit tests.  
:)